File: python/generate_key.py
"""
API Key Generator

This script will generate an API key for an Algolia application.

The generated key will be valid for Search operations, and will be limited to 100 queries per hour.

"""

from os import getenv

# Install the API client: https://www.algolia.com/doc/api-client/getting-started/install/python/?client=python
from algoliasearch.search_client import SearchClient
from dotenv import find_dotenv, load_dotenv
import json

load_dotenv(find_dotenv())

# Get your Algolia Application ID and (admin) API key from the dashboard: https://www.algolia.com/account/api-keys
# and choose a name for your index. Add these environment variables to a `.env` file:
ALGOLIA_APP_ID = getenv('ALGOLIA_APP_ID')
ALGOLIA_API_KEY = getenv('ALGOLIA_API_KEY')
ALGOLIA_INDEX_NAME = getenv('ALGOLIA_INDEX_NAME')

# Start the API client
# https://www.algolia.com/doc/api-client/getting-started/instantiate-client-index/
client = SearchClient.create(ALGOLIA_APP_ID, ALGOLIA_API_KEY)

# Create the key
#https://www.algolia.com/doc/api-reference/api-methods/add-api-key/?client=python

# Set the permissions for the key
acl = ['search']

# Set the parameters for the key
params = {
    'description': 'Restricted search-only API key for algolia.com',
    # Rate-limit to 100 requests per hour per IP address
    'maxQueriesPerIPPerHour': 100
}

# Make the API request to add the key
try:
    res = client.add_api_key(acl, params)
    key = res['key']
except Exception as e:
    print(f'Error: {e}')
    exit()

# Initialise a new client with the generated key
client = SearchClient.create(ALGOLIA_APP_ID, key)

# Create an index (or connect to it, if an index with the name `ALGOLIA_INDEX_NAME` already exists)
# https://www.algolia.com/doc/api-client/getting-started/instantiate-client-index/#initialize-an-index
index = client.init_index(ALGOLIA_INDEX_NAME)

# Test the created key
try:
    res = index.search('')
    if res:
        print(f'Key generated successfully: {key}')
except Exception as e:
    print(f'Error: {e}')
    exit()

File: php/generate_key.php
<?php

#API Key Generator
#This script will generate an API key for an Algolia application.
#The generated key will be valid for Search operations, and will be limited to 100 queries per hour.

# Install the API client: https://www.algolia.com/doc/api-client/getting-started/install/php/?client=php
require __DIR__.'/vendor/autoload.php';

use Algolia\AlgoliaSearch\Api\SearchClient;

$dotenv = Dotenv\Dotenv::createImmutable(__DIR__);
$dotenv->load();

# Algolia client credentials
$ALGOLIA_APP_ID = $_ENV['ALGOLIA_APP_ID'];
$ALGOLIA_API_KEY = $_ENV['ALGOLIA_API_KEY'];
$ALGOLIA_INDEX_NAME = $_ENV['ALGOLIA_INDEX_NAME'];

# Initialize the client
# https://www.algolia.com/doc/api-client/getting-started/instantiate-client-index/
$client = SearchClient::create($ALGOLIA_APP_ID, $ALGOLIA_API_KEY);

# Set permissions for API key
# https://www.algolia.com/doc/api-reference/api-methods/add-api-key/#method-param-acl

//Set the parameters for API key
//https://www.algolia.com/doc/rest-api/search/add-api-key

$acl = [
    'acl' => [
        'search',

        'addObject',
    ],
        'description' => 'Restricted search-only API key for algolia.com',
        'maxQueriesPerIPPerHour' => 100
    ];

# Create a new restricted search-only API key
print("Creating new key...\n");

$res = $client->addApiKey($acl);

$client->waitForTask($ALGOLIA_INDEX_NAME, $res['key']);

$new_key = $res['key'];

if ($new_key = $res['key']) {
    echo "Key generated successfully: $new_key \n";
} else {
    echo "Error while creating key\n";
}

//Wait for the API key to be created
$res = $client->waitForApiKey(
    $new_key,
    'add',
);

# Test the created key
print("Testing key...\n");

# Initialise a new client with the generated key
$newClient = SearchClient::create($ALGOLIA_APP_ID, $new_key);

# Test the new generated key by performing a search

$searchQuery =  [
    'requests' => [
        ['indexName' => $ALGOLIA_INDEX_NAME,
            'query' => '',
            'hitsPerPage' => 50,
        ],
    ],
];

if ($search_res = $newClient->search($searchQuery)) {
    echo "Successful key test\n";
} else {
    echo "Failed search with the new key\n";
}